src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IAsyncEnumerableOfTConverter.cs

```diff
@@ -59,6 +59,16 @@ protected override bool OnWriteResume(Utf8JsonWriter writer, TAsyncEnumerable va
                 // ensure the enumerator already is stored
                 // in the WriteStack for proper disposal.
                 moveNextTask = enumerator.MoveNextAsync();
+
+                if (!moveNextTask.IsCompleted)
+                {
+                    // It is common for first-time MoveNextAsync() calls to return pending tasks,
+                    // since typically that is when underlying network connections are being established.
+                    // For this case only, suppress flushing the current buffer contents (e.g. the leading '[' token of the written array)
+                    // to give the stream owner the ability to recover in case of a connection error.
+                    state.SuppressFlush = true;
+                    goto SuspendDueToPendingTask;
```

Comment:
can we avoid the goto? i.e. add bool flag which causes the loop logic to be skipped? I think saving few instructions of CPU cycle is not worth having goto in the code

Reply:
Removing the `goto` in this case would make flow substantially more complex. The only viable alternative in my mind is to use a local method, but that comes with overhead. This particular labeling strategy (inserting a label at the tail end of a method to signify the "default" return mode) is fairly similar to tail calling into a local method, so I believe it doesn't impact legibility substantially.

We actually do this quite a lot in STJ, the product code currently has ~150 `goto` statements.